### PR TITLE
Force purge if deployment status is not found when deployments are li…

### LINKED ISSUE #238: Unable to purge an application that appears in the list
# Issue Type

Put an X between the brackets of the corresponding issue type

* [X] Bug report
* [ ] Documentation issue report
* [ ] Improvement request

## Description
Unable to delete an application that appears in the "yorc d list" command.

## Expected behavior
We should be able to do that without issue even the state of this deployment is strange.

## Actual behavior

```
[root@node1 yorc]# yorc d list
Deployments:
+--------------------------------------------------------+------------+
| Id                                                     | Status     |
+--------------------------------------------------------+------------+
| FMLE__6e2d182d_5f8e_4db1_b9d0_99e56505cc1f-Environment | UNDEPLOYED |
+--------------------------------------------------------+------------+

[root@node1 yorc]# yorc d undeploy --purge FMLE__6e2d182d_5f8e_4db1_b9d0_99e56505cc1f-Environment
The deployment with the following id "FMLE__6e2d182d_5f8e_4db1_b9d0_99e56505cc1f-Environment" doesn't exist

[root@node1 yorc]# curl '127.0.0.1:8500/v1/kv/_yorc/deployments/FMLE__6e2d182d_5f8e_4db1_b9d0_99e56505cc1f-Environment/workflows/install/steps/Compute_install/status?raw&pretty'
ERROR

[root@node1 yorc]# curl '127.0.0.1:8500/v1/kv/_yorc/deployments/?keys&pretty'
[
    "_yorc/deployments/FMLE__6e2d182d_5f8e_4db1_b9d0_99e56505cc1f-Environment/workflows/install/steps/Compute_install/status"
]


[root@node1 yorc]# curl '127.0.0.1:8500/v1/kv/_yorc/?keys&pretty'
[
    "_yorc/database_schema_version",
    "_yorc/deployments/FMLE__6e2d182d_5f8e_4db1_b9d0_99e56505cc1f-Environment/workflows/install/steps/Compute_install/status",
    "_yorc/events/FMLE_Monitoring__3abaf826_efab_401b_a99e_9cfb311ae38d-Environment/2018-12-17T09:57:23.315409579Z",
    "_yorc/events/FMLE__131d6c25_f42d_4e21_9c51_4ff4471a32ce-Environment/2018-12-17T09:57:19.208117444Z",
    "_yorc/events/FMLE__2ebad9fc_afe9_4f0c_ac97_1deaf474e76c-Environment/2018-12-17T09:56:49.383969299Z",
    "_yorc/events/FMLE__31cb1907_e4e4_4f7d_b6ec_8485cdd3a8dc-Environment/2018-12-17T09:57:41.749891874Z",
    "_yorc/events/FMLE__31cb1907_e4e4_4f7d_b6ec_8485cdd3a8dc-Environment/2018-12-17T09:57:41.783387408Z",
    "_yorc/events/FMLE__4c15f3c2_4bcf_4540_af05_f5963f1085de-Environment/2018-12-17T09:56:50.784960557Z",
    "_yorc/events/FMLE__4c15f3c2_4bcf_4540_af05_f5963f1085de-Environment/2018-12-17T09:56:50.885715972Z",
    "_yorc/events/FMLE__85e16294_1de7_46ce_b249_1495361c34da-Environment/2018-12-17T09:56:54.807703335Z",
    "_yorc/events/FMLE__b304e4b3_27e4_4152_9b98_e8cc602fba3a-Environment/2018-12-17T09:56:57.56943443Z",
    "_yorc/events/FMLE__b304e4b3_27e4_4152_9b98_e8cc602fba3a-Environment/2018-12-17T09:56:57.590276231Z",
    "_yorc/events/FMLE__ba5576f5_8d3f_4e1c_b5a9_551c50d2b18a-Environment/2018-12-17T09:56:58.770599027Z",
    "_yorc/events/FMLE__ba5576f5_8d3f_4e1c_b5a9_551c50d2b18a-Environment/2018-12-17T09:56:58.820478333Z",
    "_yorc/events/FMLE__c0db0efd_f1a5_400a_b83b_d83ed2e5f88e-Environment/2018-12-17T09:56:59.769970272Z",
    "_yorc/events/FMLE__c0db0efd_f1a5_400a_b83b_d83ed2e5f88e-Environment/2018-12-17T09:56:59.80488674Z",
    "_yorc/events/FMLE__c61e36ee_24cd_4e44_b3d3_bd43ade06537-Environment/2018-12-17T09:57:00.003010914Z",
    "_yorc/events/FMLE__c61e36ee_24cd_4e44_b3d3_bd43ade06537-Environment/2018-12-17T09:57:00.046812653Z",
    "_yorc/events/FMLE__c72022f2_5aca_408e_a656_600ed85f06cf-Environment/2018-12-17T09:57:00.264499257Z",
    "_yorc/events/FMLE__c72022f2_5aca_408e_a656_600ed85f06cf-Environment/2018-12-17T09:57:00.294245763Z",
    "_yorc/events/FMLE__c95954c3_762b_4104_9cd0_08aad2aee95e-Environment/2018-12-17T09:57:42.476224203Z",
    "_yorc/events/FMLE__cfbd29ac_f8e5_4216_8989_64fee6a8f6e4-Environment/2018-12-17T09:57:00.966894244Z",
    "_yorc/events/FMLE__cfbd29ac_f8e5_4216_8989_64fee6a8f6e4-Environment/2018-12-17T09:57:01.017712538Z",
    "_yorc/events/FMLE__d2668b43_cd6d_4301_bd15_85b8139269d3-Environment/2018-12-17T09:57:01.618918677Z",
    "_yorc/events/FMLE__d2668b43_cd6d_4301_bd15_85b8139269d3-Environment/2018-12-17T09:57:01.660957323Z",
    "_yorc/events/FMLE__d3204454_d7b1_4d5c_8b28_87431bba18f3-Environment/2018-12-17T09:57:01.860208523Z",
    "_yorc/events/FMLE__d6c16539_6d25_4cb7_93d7_fcbbaf15b0ad-Environment/2018-12-17T09:57:02.17133087Z",
    "_yorc/events/FMLE__d6c16539_6d25_4cb7_93d7_fcbbaf15b0ad-Environment/2018-12-17T09:57:02.199526805Z",
    "_yorc/events/FMLE__de6868ab_cc6c_463c_ab88_2d323367cb56-Environment/2018-12-17T09:57:02.692328169Z",
    "_yorc/events/FMLE__ea9df519_5696_401c_a855_91d192410588-Environment/2018-12-17T09:57:03.311773819Z",
    "_yorc/events/FMLE__f9d313dd_d719_4f08_82e0_11700b14a0e2-Environment/2018-12-17T09:57:03.860920395Z",
    "_yorc/events/FMLE__fb566c84_12bd_458d_847a_656aa174ff25-Environment/2018-12-17T09:57:31.571744817Z",
    "_yorc/events/FMLE__fd965dd7_7bcd_446a_bb70_03caa2795e0e-Environment/2018-12-17T09:57:04.660709724Z",
    "_yorc/events/FMLE__fd965dd7_7bcd_446a_bb70_03caa2795e0e-Environment/2018-12-17T09:57:04.684788255Z",
    "_yorc/executions/.processingLock-d0a237c0-1eda-4e13-8b0c-967eecca16f2",
    "_yorc/executions/.processingLock-ecd57e22-57d1-4a83-a839-898f5bdd74ad",
    "_yorc/hosts_pool/10.197.138.173/connection/host",
    "_yorc/hosts_pool/10.197.138.173/connection/password",
    "_yorc/hosts_pool/10.197.138.173/connection/port",
    "_yorc/hosts_pool/10.197.138.173/connection/private_key",
    "_yorc/hosts_pool/10.197.138.173/connection/user",
    "_yorc/hosts_pool/10.197.138.173/labels/environment",
    "_yorc/hosts_pool/10.197.138.173/labels/public_address",
    "_yorc/hosts_pool/10.197.138.173/labels/testlabel",
    "_yorc/hosts_pool/10.197.138.173/message",
    "_yorc/hosts_pool/10.197.138.173/status",
    "_yorc/logs/FMLE_Monitoring__3abaf826_efab_401b_a99e_9cfb311ae38d-Environment/2018-12-17T09:57:23.320956301Z",
    "_yorc/logs/FMLE__131d6c25_f42d_4e21_9c51_4ff4471a32ce-Environment/2018-12-17T09:57:19.212382725Z",
    "_yorc/logs/FMLE__2ebad9fc_afe9_4f0c_ac97_1deaf474e76c-Environment/2018-12-17T09:56:49.404314889Z",
    "_yorc/logs/FMLE__31cb1907_e4e4_4f7d_b6ec_8485cdd3a8dc-Environment/2018-12-17T09:57:41.76050227Z",
    "_yorc/logs/FMLE__31cb1907_e4e4_4f7d_b6ec_8485cdd3a8dc-Environment/2018-12-17T09:57:41.794745503Z",
    "_yorc/logs/FMLE__4c15f3c2_4bcf_4540_af05_f5963f1085de-Environment/2018-12-17T09:56:50.814416284Z",
    "_yorc/logs/FMLE__4c15f3c2_4bcf_4540_af05_f5963f1085de-Environment/2018-12-17T09:56:50.957870931Z",
    "_yorc/logs/FMLE__6e2d182d_5f8e_4db1_b9d0_99e56505cc1f-Environment/2018-12-14T15:36:53.292620265Z",
    "_yorc/logs/FMLE__6e2d182d_5f8e_4db1_b9d0_99e56505cc1f-Environment/2018-12-14T15:36:53.305948576Z",
    "_yorc/logs/FMLE__6e2d182d_5f8e_4db1_b9d0_99e56505cc1f-Environment/2018-12-14T15:36:53.320172838Z",
    "_yorc/logs/FMLE__6e2d182d_5f8e_4db1_b9d0_99e56505cc1f-Environment/2018-12-14T15:36:53.338757899Z",
    "_yorc/logs/FMLE__6e2d182d_5f8e_4db1_b9d0_99e56505cc1f-Environment/2018-12-14T15:36:53.343400848Z",
    "_yorc/logs/FMLE__6e2d182d_5f8e_4db1_b9d0_99e56505cc1f-Environment/2018-12-14T15:36:53.348391126Z",
    "_yorc/logs/FMLE__85e16294_1de7_46ce_b249_1495361c34da-Environment/2018-12-17T09:56:54.82893979Z",
    "_yorc/logs/FMLE__b304e4b3_27e4_4152_9b98_e8cc602fba3a-Environment/2018-12-17T09:56:57.58297698Z",
    "_yorc/logs/FMLE__b304e4b3_27e4_4152_9b98_e8cc602fba3a-Environment/2018-12-17T09:56:57.598961614Z",
    "_yorc/logs/FMLE__ba5576f5_8d3f_4e1c_b5a9_551c50d2b18a-Environment/2018-12-17T09:56:58.789423129Z",
    "_yorc/logs/FMLE__ba5576f5_8d3f_4e1c_b5a9_551c50d2b18a-Environment/2018-12-17T09:56:58.840421481Z",
    "_yorc/logs/FMLE__c0db0efd_f1a5_400a_b83b_d83ed2e5f88e-Environment/2018-12-17T09:56:59.780470448Z",
    "_yorc/logs/FMLE__c0db0efd_f1a5_400a_b83b_d83ed2e5f88e-Environment/2018-12-17T09:56:59.811845072Z",
    "_yorc/logs/FMLE__c61e36ee_24cd_4e44_b3d3_bd43ade06537-Environment/2018-12-17T09:57:00.007527828Z",
    "_yorc/logs/FMLE__c61e36ee_24cd_4e44_b3d3_bd43ade06537-Environment/2018-12-17T09:57:00.056839291Z",
    "_yorc/logs/FMLE__c72022f2_5aca_408e_a656_600ed85f06cf-Environment/2018-12-17T09:57:00.27810511Z",
    "_yorc/logs/FMLE__c72022f2_5aca_408e_a656_600ed85f06cf-Environment/2018-12-17T09:57:00.306385192Z",
    "_yorc/logs/FMLE__c95954c3_762b_4104_9cd0_08aad2aee95e-Environment/2018-12-17T09:57:42.493375532Z",
    "_yorc/logs/FMLE__cfbd29ac_f8e5_4216_8989_64fee6a8f6e4-Environment/2018-12-17T09:57:00.97588129Z",
    "_yorc/logs/FMLE__cfbd29ac_f8e5_4216_8989_64fee6a8f6e4-Environment/2018-12-17T09:57:01.024575853Z",
    "_yorc/logs/FMLE__d2668b43_cd6d_4301_bd15_85b8139269d3-Environment/2018-12-17T09:57:01.635938419Z",
    "_yorc/logs/FMLE__d2668b43_cd6d_4301_bd15_85b8139269d3-Environment/2018-12-17T09:57:01.674669904Z",
    "_yorc/logs/FMLE__d6c16539_6d25_4cb7_93d7_fcbbaf15b0ad-Environment/2018-12-17T09:57:02.192342086Z",
    "_yorc/logs/FMLE__d6c16539_6d25_4cb7_93d7_fcbbaf15b0ad-Environment/2018-12-17T09:57:02.207010262Z",
    "_yorc/logs/FMLE__ea9df519_5696_401c_a855_91d192410588-Environment/2018-12-17T09:57:03.326808627Z",
    "_yorc/logs/FMLE__f9d313dd_d719_4f08_82e0_11700b14a0e2-Environment/2018-12-17T09:57:03.891802186Z",
    "_yorc/logs/FMLE__fb566c84_12bd_458d_847a_656aa174ff25-Environment/2018-12-17T09:57:31.576755324Z",
    "_yorc/logs/FMLE__fd965dd7_7bcd_446a_bb70_03caa2795e0e-Environment/2018-12-17T09:57:04.667335754Z",
    "_yorc/logs/FMLE__fd965dd7_7bcd_446a_bb70_03caa2795e0e-Environment/2018-12-17T09:57:04.691226765Z",
    "_yorc/service/monitoring/leader",
    "_yorc/service/scheduling/leader",
    "_yorc/workflows/48fc1ce2-ebcb-4b24-905d-71ba468b2592/Compute_install"
]

[root@node1 yorc]# curl '127.0.0.1:8500/v1/kv/_yorc/deployments/?keys&pretty'
[
    "_yorc/deployments/FMLE__6e2d182d_5f8e_4db1_b9d0_99e56505cc1f-Environment/workflows/install/steps/Compute_install/status"
]

```

## Steps to reproduce the issue

Difficult to reproduce

## Yorc configuration file

```YAML
#use_ssh_agent: false
workers_number: 50

ansible:
    use_openssh: true

infrastructures:
    slurm:
        url: 10.0.0.75
        user_name: root
        private_key: /var/yorc/.ssh/yorc.pem
        port: 22

```

## Priority

**High**

* High = We should stop anything we are doing and do this. If you're writing this calmly, you should probably choose another option. If it's really that urgent, please make sure we read this.
* Medium = It's important that we do this within a few days.
* Low = We will consider this on our next sprint planning.

(Please be aware that your priority may not match ours, we'll use this as guidance only).

### PR DESCRIPTION
**Pull request by stebenoist**
# Pull Request description

## Description of the change
- Ensure purge is done if requested with undeploy
- Create taskTypeForcePurge to force deployment removal without any status check
- Force purge if deployment status is not found when deployments are listed and do not return it

### How to verify it
Deploy an app
Undeploy it with Yorc CLI without purge option (yorc d undeploy)
Remove deployment status in Consul
Check with "yorc d list" that the deployment is not listed and it has been removed from Consul

## Applicable Issues
fixes #238

### CODE REVIEW

File: tasks/workflow/worker.go [hunk 1]
```diff
@@ -216,13 +216,13 @@ func (w *worker) handleExecution(t *taskExecution) {
...
 		err = w.runCustomWorkflow(ctx, t, wfName)
-
 	case tasks.TaskTypeAction:
 		err = w.runAction(ctx, t)
-
 	case tasks.TaskTypeQuery, tasks.TaskTypeCustomCommand:
 		// Those kind of task will manage monitoring of taskFailure differently
 		err = w.runOneExecutionTask(ctx, t)
+	case tasks.TaskTypeForcePurge:
+		err = w.runPurge(ctx, t)
```
[the remark this answers is not in the document]
Comment (stebenoist) [reply tasks/workflow/worker.go]: Yes, let's do this !

File: tasks/workflow/worker.go [hunk 2]
```diff
@@ -580,20 +587,11 @@ func (w *worker) runUndeploy(ctx context.Context, t *taskExecution) error {
...
 				return err
 			}
-			if t.taskType == tasks.TaskTypePurge {
-				err = w.runPurge(ctx, t)
-				if err != nil {
-					return err
-				}
-			}
-			return nil
+			return err
```
[the remark this answers is not in the document]
Comment (stebenoist) [reply tasks/workflow/worker.go]: Need to handle potentially multiple errors

File: tasks/workflow/worker.go [hunk 3]
```diff
@@ -571,23 +572,27 @@ func (w *worker) runUndeploy(ctx context.Context, t *taskExecution) error {
...
+					if err != nil {
+						mErr = multierror.Append(mErr, err)
+					}
 				}
 			}()
-			_, err = updateTaskStatusAccordingToWorkflowStatus(ctx, t.cc.KV(), t.targetID, t.taskID, "uninstall")
+			_, err := updateTaskStatusAccordingToWorkflowStatus(ctx, t.cc.KV(), t.targetID, t.taskID, "uninstall")
 			if err != nil {
-				return err
+				return multierror.Append(mErr, err)
```
[the remark this answers is not in the document]
Comment (stebenoist) [reply tasks/workflow/worker.go]: I'm afraid it doesn't.

[the remark this answers is not in the document]
Comment (stebenoist) [reply tasks/workflow/worker.go]: So, I made it simpler !

[the remark this answers is not in the document]
Comment (loicalbertin) [reply tasks/workflow/worker.go]: That said, having such a functionality to accumulate errors and fail later is interesting.
We just have to make it a way it honor standard go formatting properly.

**Pull request merged**